asset/common.go

```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		return nil, err
+	}
+	AssetStore = &GlobalAssetStore{
+		DB: globalAssetDB,
+		FS: &SwiftFS{Connection: sc},
+	}
+	if err != nil {
+		return nil, err
+	}
+	return AssetStore, nil
+}
+
+// MarshalAssetKey returns the string key store in UsedBy field for app versions
+func MarshalAssetKey(spacePrefix, appSlug, version string) string {
+	if spacePrefix == consts.DefaultSpacePrefix {
+		spacePrefix = ""
+	}
+	return filepath.Join(spacePrefix, appSlug, version)
+}
+
+func InitCouchDB(addr, user, pass, prefix string) (*kivik.DB, error) {
+	u, err := url.Parse(addr)
+	if err != nil {
+		return nil, err
+	}
+	u.User = nil
+
+	client, err = kivik.New("couch", u.String())
+	if err != nil {
+		return nil, err
+	}
+
+	if user != "" {
+		err = client.Authenticate(ctx, &chttp.BasicAuth{
+			Username: user,
+			Password: pass,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	assetsStoreDBName := "registry-" + assetStoreDBSuffix
+	exists, err := client.DBExists(ctx, assetsStoreDBName)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		fmt.Printf("Creating database %q...", assetsStoreDBName)
+		if err := client.CreateDB(ctx, assetsStoreDBName); err != nil {
+			return nil, err
+		}
+		fmt.Println("ok.")
+	}
+
+	globalAssetStoreDB := client.DB(ctx, assetsStoreDBName)
+	if err = globalAssetStoreDB.Err(); err != nil {
+		return nil, err
+	}
+
+	return globalAssetStoreDB, nil
+}
+
+func InitSwift() (*swift.Connection, error) {
+	conf, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	sc := conf.SwiftConnection
+
+	if err := sc.ContainerCreate(AssetContainerName, nil); err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+func (a *GlobalAssetStore) AddAsset(asset *GlobalAsset, content io.Reader, source string) error {
+	// Creating the asset in the FS
+	err := a.FS.AddAsset(asset, content)
+	if err != nil {
+		return err
+	}
+
+	// Handles the CouchDB updates
+	var doc *GlobalAsset
+	row := AssetStore.DB.Get(ctx, asset.MD5, nil)
+	err = row.ScanDoc(&doc)
+	if err != nil && kivik.StatusCode(err) != kivik.StatusNotFound {
+		return err
+	}
+
+	var docRev string
+	// If asset does not exist in CouchDB global asset database, create it
+	if kivik.StatusCode(err) == kivik.StatusNotFound {
+		doc = asset
+		doc.ID = asset.MD5
+		_, docRev, err = AssetStore.DB.CreateDoc(ctx, doc)
+		if err != nil {
+			return err
+		}
+		doc.Rev = docRev
+	}
+
+	// Updating the UsedBy field to add the new app version
+	found := false
+	for _, usedBy := range doc.UsedBy {
+		if usedBy == source {
+			found = true
+			break
+		}
+	}
+	if !found {
+		doc.UsedBy = append(doc.UsedBy, source)
+	}
+	_, err = AssetStore.DB.Put(ctx, doc.ID, doc, nil)
+	return err
+}
+
+func (a *GlobalAssetStore) RemoveAsset(md5, versionFilepath string) error {
+	row := AssetStore.DB.Get(ctx, md5)
+
+	var assetDoc *GlobalAsset
+	err := row.ScanDoc(&assetDoc)
+	if err != nil && kivik.StatusCode(err) != kivik.StatusNotFound {
+		return err
```

Comment:
If `kivik.StatusCode(err) == kivik.StatusNotFound`, `assetDoc` will be nil and it will panic later: we should also return `err` in this case:

```go
if err != nil {
    return err
}
```